numpy/_core/src/multiarray/alloc.cpp

```diff
@@ -50,8 +50,26 @@ typedef struct {
     npy_uintp available; /* number of cached pointers */
     void * ptrs[NCACHE];
 } cache_bucket;
-static cache_bucket datacache[NBUCKETS];
-static cache_bucket dimcache[NBUCKETS_DIM];
+
+static thread_local cache_bucket datacache[NBUCKETS];
+static thread_local cache_bucket dimcache[NBUCKETS_DIM];
```

Comment:
I guess now that we're requiring C11 atomics we can replace all internal uses of `NPY_TLS` with `thread_local`. Not needed to fix that here, just noting that because I noticed this doesn't use `NPY_TLS`.